File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    bool go(TreeNode* root, TreeNode* subroot){
        if(!root and !subroot) return true;
        if(!root or !subroot or root->val !=subroot->val) return false;
        return go(root->left, subroot->left) and go(root->right,subroot->right);
    }
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        if(!subRoot) return true;
        if(!root) return false;
        return go(root,subRoot) or isSubtree(root->left,subRoot) or isSubtree(root->right,subRoot);
    }
};

File: Data Structures & Algorithms/balanced-binary-tree/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    
    def height(self,root):
        if not root:
            return 0
        lh, rh = self.height(root.left), self.height(root.right)
        if abs(lh-rh)>1:
            self.flag = False
        return max(lh,rh)+1

    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        self.flag=True
        self.height(root)
        return self.flag